### DISCUSSION TITLE
Not retrieving all (sub)categories

### DISCUSSION THREAD

**Discussion by Kealtie** [swellstores/community#502] [Help] [answered]
Hello,

In the application we're developing for our customer we have a navigation menu, containing all categories and (sub)sub-categories. This is a three level menu, on hover it displays the sub-categories (child-categories) of the currently hovered category.

We've written logic that assigns the sub-categories to a parent, which works. But i've noticed we just don't receive all active categories from the API.

For example, we've got the following setup in Swell.
<img width="1127" alt="image" src="https://github.com/swellstores/community/assets/7847482/e846e49c-8c9b-495b-abcd-34d04886cd4e">

As you can see they're all active categories. But we do NOT get the category with the name: C-27/60 in our GET request for all categories. 

When I switch positions between U-27 and C-27/60, I do retrieve C-27/60, which is weird.

Here's the place where we do our request(s). 
```
// Fetch categories from Swell API and loop through all pages
    const {
              data   : categoryResult,
              pending: isLoadingCategories,
              error,
              refresh: refreshCategories,
          } = await useSingletonAsyncData<SwellCategory|null>('categories', async () => {
        let allCategories: SwellCategory[] = [];
        let hasMorePages = true;

        // Loop through all pages
        while (hasMorePages) {
            const result = await useSwellApi('/categories', {
                params: {
                    limit: 100,
                    page : currentPage.value
                },
            });

            // Handle errors
            if (result.error) {
                alert('danger', {
                    title    : t('alerts.fetching-categories-failed'),
                    countdown: 4000,
                });
                break;
            }

            const currentCategories = result.results ?? [];
            allCategories = [...allCategories, ...currentCategories];

            // Check if there are more pages, if not, reset to page 1
            if (currentPage.value === result.pageCount) {
                hasMorePages = false;
                currentPage.value = 1;
            } else {
                currentPage.value++;
            }
        }

        return { results: allCategories };
    });
```

Since the API limits us from retrieving more then 100 categories per request, we're looping the request until we've fetched the last page. 

What could this possibly be? This is just an example category, but this happens with more categories. 

I've used the Developer Console in the Swell UI, I do retrieve all categories there, when I do not filter on the ones that are 'active'. But when I do, I retrieve the exact same result.

**Comment by logeshswell** [comment]
Can you share your store link along with the screenshot from **Developer console**?
Let me also know the category link that you are not seeing in the response.

**Reply by Kealtie** [reply]
This would be the store link: https://obimex-bv.swell.store/admin.
This would be one of the categories that is not fetching. https://obimex-bv.swell.store/admin/categories/659ba788fe54920012e247da

**Reply by logeshswell** [reply]
I just ran a sample query to fetch active categories with limit: 500 to search for the record faster.
It returned the category that you shared.

<img width="1229" alt="Screenshot 2024-02-20 at 1 09 23 PM" src="https://github.com/swellstores/community/assets/100766131/e5c6b153-91a4-4fcd-bc48-d9b165a6f520">

<img width="587" alt="Screenshot 2024-02-20 at 1 07 01 PM" src="https://github.com/swellstores/community/assets/100766131/36d09da4-3c57-428b-a3db-cf7aa13f840e">

**Reply by logeshswell** [reply]
Can you check and confirm please?

**Reply by Kealtie** [reply]
Yes, but as stated in my initial post. I do not retrieve those categories through my code, also I can not set my limit to 500 as it is capped at 100. 

It's weird cause at the time writing the initial post, I wouldn't retrieve the sub-category of metal-stud-profielen.

![image](https://github.com/swellstores/community/assets/7847482/fd4a34f0-f3b9-415e-9ef2-abb91b5ae02e)


This is a screenshot of the navigation that is created by categories & sub-categories. 
Within the fetched categories I can not seem to find U-27. 

![image](https://github.com/swellstores/community/assets/7847482/64ba586e-e399-4ded-a909-0bf9e4aa9ac6)

**Reply by logeshswell** [reply]
> I do not retrieve those categories through my code

1. Can you share your code snippet that calls Swell sdk alone?
2. Do you use swell-js or swell-node?

**Reply by Kealtie** [reply]
We're creating a $fetch instance like so:

```
import { camelizeKeys } from 'fast-case';
import { NitroFetchOptions, NitroFetchRequest } from 'nitropack';

/**
 * Converts a string to base64
 * @param str
 */
function base64Encode (str: string): string {
    if (typeof window !== 'undefined') {
        return window.btoa(str);
    }
    return Buffer.from(str).toString('base64');
}

/**
 * Creates an OhMyFetch instance with the Swell API base URL and headers.
 * This instance can be used to make requests to the Swell API.
 * @param url
 * @param options
 */
export default function <ReturnType>(url: string, requestOptions: NitroFetchOptions<NitroFetchRequest> = {}) {
    let encodedPublicKey = '';
    let baseUrl = '';

    try {
        const { SWELL_STOREFRONT_API_KEY, SWELL_STORE_ID } = useRuntimeConfig();

        encodedPublicKey = base64Encode(SWELL_STOREFRONT_API_KEY);
        baseUrl = new URL(`https://${SWELL_STORE_ID}.swell.store/api`).toString();
    } catch (e) {

    }

    const swellClient = $fetch.create({
        baseURL: baseUrl,
        async onRequest ({ options }) {
            try {
                const cookieSession = useCookie('swell-session');
                const swellLocale = useCookie('swell-locale');
                const swellCurrency = useCookie('swell-currency');

                options.headers = {
                    Accept         : 'application/json',
                    'Content-Type' : 'application/json',
                    Authorization  : `Basic ${encodedPublicKey}`,
                };

                if (cookieSession.value) {
                    options.headers['X-Session'] = cookieSession.value;
                }

                if (swellLocale.value) {
                    options.headers['X-Locale'] = swellLocale.value;
                }

                if (swellCurrency.value) {
                    options.headers['X-Currency'] = swellCurrency.value;
                }
            } catch (e) {
            }
        },

        async onResponse ({ response }) {
            try {
                const responseSession = response.headers.get('X-Session');
                const cookieSession = useCookie('swell-session');
                if (typeof responseSession === 'string' && cookieSession.value !== responseSession) {
                    useCookie('swell-session').value = responseSession;
                }
            } catch (e) {
            }
        },
        parseResponse (response) {
            const responseJson = JSON.parse(response);
            return camelizeKeys(responseJson);
        },
    });

    return swellClient<ReturnType>(url, requestOptions);
}
```

**Reply by Kealtie** [reply]
We're using Nuxt 3 using the Composition API fyi.

The contents of the code above would be the contents of the useSwellApi composable.

**Reply by logeshswell** [reply]
> 2. Do you use swell-js or swell-node?

I can see that you use swell-js (or [Swell frontend API](https://developers.swell.is/frontend-api/categories))

**Reply by logeshswell** [reply]
Let me check and keep you posted shortly now. I can also reproduce this issue in your store.

**Reply by logeshswell** [reply]
Is this the only category affected in this API call?

**Reply by Kealtie** [reply]
No, multiple active categories are not in the response. 

But, I do not have a logical explanation for what the reason might be for a category not to be in the response. Therefore I can not give you exact categories that work and do not work.

**Reply by logeshswell** [reply]
No, if you share another 1 or 2 active category ID that are affected, it would be very helpful to debug this issue faster and make a fix for it.

**Reply by logeshswell** [reply]
No thanks, I can confirm that there's a bug present in fetching the active categories via swell frontend API.
We'll take a deeper look at this issue and get back to you if I get some updates.

**Reply by Kealtie** [reply]
Yes, alright. For instance, we have two 'Metal Stud Profielen' categories, one is a sub-category of 'Metal Stud Plafonds', which is the one we've described in the above.

But also the 'Metal Stud Profielen' which is a sub-category of 'Metal Stud Wanden' (see screenshot below) does NOT show any of it's subcategories in our navigation. 
![image](https://github.com/swellstores/community/assets/7847482/9cdd496b-40d9-46a2-abd0-996393cb4818)

Now i've tried renaming one of the two 'Metal Stud Profielen' categories, as I thought the same name might be the issue here, but that did not resolve the issue.

**Reply by logeshswell** [reply]
Yeah, I found that 22 such categories were not returned in the frontend API.

**Reply by Kealtie** [reply]
Thanks, looking forward to your response.

**Comment by logeshswell** [comment, marked as answer]
There was a bug with the pagination and it is fixed now. Can you please check and confirm if it's working fine?
